### DISCUSSION TITLE
Form validation on a QDialog not working

### DISCUSSION THREAD

**Discussion by eugenevk** [quasarframework/quasar#12205] [General - Components / Directives / etc] [answered]
I want to validate the fields on a dialog and have tried the QInput validation example from the documentation. As soon as I click the mouse, the dialog is closed and no validation happens. What am I missing?
My dialog template:
```
<template>
  <q-page class="flex flex-center">
    <q-btn @click="dialog = true">Open dialog</q-btn>
    <q-dialog v-model="dialog">
      <form @submit.prevent.stop="onSubmit">
        <q-card class="col" style="width: 400px; max-width: 500px">
          <q-card-section align="center"> Details </q-card-section>
          <q-card-section class="row q-col-gutter-sm">
            <q-input
              label="First name"
              class="col-6"
              outlined
              stack-label
              v-model="firstname"
            />
            <q-input
              ref="refLastName"
              label="Last name"
              class="col-6"
              outlined
              stack-label
              :rules="[(val) => (val && val.length > 0) || 'Field is required']"
              v-model="lastname"
            >
              <template #label>
                Last name <span class="text-weight-bold text-red">*</span>
              </template>
            </q-input>
          </q-card-section>
          <q-card-actions align="center">
            <q-btn type="submit">Submit</q-btn>
            <q-btn label="Close" v-close-popup />
          </q-card-actions>
        </q-card>
      </form>
    </q-dialog>
  </q-page>
</template>
```
A codepen is here: https://codesandbox.io/s/smoosh-cherry-92k7k?file=/src/pages/Index.vue:0-1263

**Comment by yusufkandemir** [comment, marked as answer]
https://quasar.dev/vue-components/dialog#usage
> **Note**
>
> It’s best that your QDialog main content is a QCard. However, if you are planning on using any other component (like QForm) or tag, make sure that the direct child of QDialog is rendered with a \<div\> tag (or wrap it with one yourself).

You have `<form>` as a direct child of `<q-dialog>`, which messes up some stuff and results in the dialog being closed when clicked. As per the note, you should move the `<form>` tag inside `<q-card>` on your example, and the problem will be solved.

Judging from the submitting logic, and your other discussion thread #12206, I would say you probably want to use QForm, as I also stated [here](https://github.com/quasarframework/quasar/discussions/12206#discussioncomment-2054307).

**Reply by eugenevk** [reply]
Thanks @yusufkandemir, this works indeed. Next thing is that I have my own custom component for QInput. In that case the validation doesn't kick-in. This is my custom component InputText.vue:
```
<template>
  <q-input
    hide-bottom-space
    :label="label"
    :model-value="modelValue"
    outlined
    :rules="[(val) => (val && val.length > 0) || 'Field is required']"
    stack-label
    @update:model-value="$emit('update:model-value', $event)"
  >
    <template #label>
      {{ label }} <span class="text-weight-bold text-red">*</span>
    </template>
  </q-input>
</template>

<script>
export default {
  name: "InputText",
  emits: ["update:model-value"],
  props: ["label", "modelValue"],
  setup() {
    return {};
  },
};
</script>
```
And this is how I now have my parent component:
```
<template>
  <q-page class="flex flex-center">
    <q-btn @click="dialog = true">Open dialog</q-btn>
    <q-dialog v-model="dialog">
      <q-card class="col" style="width: 400px; max-width: 500px">
        <form @submit.prevent.stop="onSubmit">
          <q-card-section align="center"> Details </q-card-section>
          <q-card-section class="row q-col-gutter-sm">
            <q-input
              label="First name"
              class="col-6"
              outlined
              stack-label
              v-model="firstname"
            />
            <InputText
              ref="refLastName"
              :label="'Last name'"
              v-model="lastname"
            />
          </q-card-section>
          <q-card-actions align="center">
            <q-btn type="submit">Submit</q-btn>
            <q-btn label="Close" v-close-popup />
          </q-card-actions>
        </form>
      </q-card>
    </q-dialog>
  </q-page>
</template>

<script>
import { defineComponent, ref } from "vue";
import InputText from "components/InputText.vue";

export default defineComponent({
  name: "Index",
  components: {
    InputText,
  },
  setup() {
    const dialog = ref(false);
    const firstname = ref(null);
    const lastname = ref(null);
    const refLastName = ref(null);
    const onSubmit = () => {
      refLastName.value.validate();
      if (refLastName.value.hasError) {
        alert("Error");
      }
    };
    return {
      dialog,
      firstname,
      lastname,
      onSubmit,
      refLastName,
    };
  },
});
</script>
```
How can I activate the validation now?

Codepen is here: https://codesandbox.io/s/smoosh-cherry-92k7k?file=/src/pages/Index.vue:0-1554

**Reply by yusufkandemir** [reply]
That is expected, `refLastName` is a ref to your custom component, not directly to q-input. And, your custom component doesn't have `validate()` or `hasError`, so you would need to expose and proxy the related methods/properties from your custom input component to call the method from the QInput.
You can avoid all these by using QForm as I stated earlier. If you are going to ignore my advice about that, you can proxy the methods as I said.

**Reply by eugenevk** [reply]
I do want to use QForm (as you can see in the code above), but also custom component for QInput, because in my project I want to do numerous standard settings for QInput. So how can I make it easy with QForm?

**Reply by yusufkandemir** [reply]
Have you checked https://quasar.dev/vue-components/form#usage?
It contains all the information and examples you need. You custom component fits into this category:
> QForm hooks into QInput, QSelect, or QField wrapped components

So, you just need to update `<form @submit.prevent.stop="onSubmit">` to `<q-form @submit="onSubmit">`. `onSubmit` will only be triggered when the validation passes, so you can get rid of `refLastName` and the error checking logic.

**Reply by eugenevk** [reply]
Many thanks @yusufkandemir. Am fairly new to Quasar and trying to read and understand as much as possible, but sometimes I don't know where to find the relevant information. Your help is much appreciated!

**Reply by eugenevk** [reply]
@yusufkandemir, I have removed the refLastName, but it still doesn't work. The validation in my custom component InputText doesn't get triggered. This is my template:
```
<template>
  <q-page class="flex flex-center">
    <q-btn @click="dialog = true">Open dialog</q-btn>
    <q-dialog v-model="dialog">
      <q-card class="col" style="width: 400px; max-width: 500px">
        <form @submit="onSubmit">
          <q-card-section align="center"> Details </q-card-section>
          <q-card-section class="row q-col-gutter-sm">
            <q-input
              label="First name"
              class="col-6"
              outlined
              stack-label
              v-model="firstname"
            />
            <InputText :label="'Last name'" v-model="lastname" />
          </q-card-section>
          <q-card-actions align="center">
            <q-btn type="submit">Submit</q-btn>
            <q-btn label="Close" v-close-popup />
          </q-card-actions>
        </form>
      </q-card>
    </q-dialog>
  </q-page>
</template>

<script>
import { defineComponent, ref } from "vue";
import InputText from "components/InputText.vue";

export default defineComponent({
  name: "Index",
  components: {
    InputText,
  },
  setup() {
    const dialog = ref(false);
    const firstname = ref(null);
    const lastname = ref(null);
    const onSubmit = () => {
      alert("hi");
    };
    return {
      dialog,
      firstname,
      lastname,
      onSubmit,
    };
  },
});
</script>
```
This is the custom component:
```
<template>
  <q-input
    hide-bottom-space
    :label="label"
    lazy-rules
    :model-value="modelValue"
    outlined
    :rules="[(val) => (val && val.length > 0) || 'Field is required']"
    stack-label
    @update:model-value="$emit('update:model-value', $event)"
  >
    <template #label>
      {{ label }} <span class="text-weight-bold text-red">*</span>
    </template>
  </q-input>
</template>

<script>
export default {
  name: "InputText",
  emits: ["update:model-value"],
  props: ["label", "modelValue"],
  setup() {
    return {};
  },
};
</script>
```
The codepen is still here: https://codesandbox.io/s/smoosh-cherry-92k7k?file=/src/components/InputText.vue:0-564
Can you give it one more look please?

**Reply by yusufkandemir** [reply]
You are still using `<form>`: **FORM** -> **F** **O** **R** **M**
You need to use `<q-form>`: **Q-FORM** -> **Q** **-** **F** **O** **R** **M**

**Reply by eugenevk** [reply]
Gosh, completely overlooked. It's working! Thanks a mil @yusufkandemir
